main/ZactuatorFASTLED.ino

```diff
@@ -0,0 +1,293 @@
+long currentUpdate = 0;
+CRGB leds[FASTLED_NUM_LEDS];
+CRGB ledColorBlink[FASTLED_NUM_LEDS];
+StaticJsonBuffer<200> jsonBuffer;
+bool blinkLED[FASTLED_NUM_LEDS];
+const long blinkInterval = 300;
+const long fireUpdate = 10;
+unsigned long previousMillis = 0;
+CRGBPalette16 gPal;
+
+void setupFASTLED()
+{
+  trc(F("FASTLED_DATA_PIN "));
+  trc(String(FASTLED_DATA_PIN));
+  trc(F("FASTLED_NUM_LEDS "));
+  trc(String(FASTLED_NUM_LEDS));
+  trc(F("ZactuatorFASTLED setup done "));
+  FastLED.addLeds<FASTLED_TYPE, FASTLED_DATA_PIN>(leds, FASTLED_NUM_LEDS);
+}
+
+//returns the current step of the animation
+int animation_step(int duration, int steps)
+{
+  int currentStep = ((currentUpdate % duration) / ((float)duration)) * steps;
+  return currentStep;
+}
+
+//returns the number of steps since the last update
+int animation_step_count(int duration, int steps)
+{
+  long lastAnimationNumber = lastUpdate / duration;
+  long currentAnimationNumber = currentUpdate / duration;
+  int lastStep = ((lastUpdate % duration) / ((float)duration)) * steps;
+  int currentStep = ((currentUpdate % duration) / ((float)duration)) * steps;
+
+  return currentStep - lastStep + (currentAnimationNumber - lastAnimationNumber) * steps;
+}
+
+void FASTLEDLoop()
+{
+
+  lastUpdate = currentUpdate;
+  currentUpdate = millis();
+
+  if (currentLEDState == GENERAL)
+  {
+
+    for (int i = 0; i < FASTLED_NUM_LEDS; i++)
+    {
+
+      int count = animation_step_count(blinkInterval, 2);
+      if (count > 0)
+      {
+        int step = animation_step(blinkInterval, 2);
+
+        if (blinkLED[i])
+        {
+
+          if (step == 0)
+          {
+            leds[i] = ledColorBlink[i];
+          }
+          else
+          {
+            ledColorBlink[i] = leds[i];
+            leds[i] = CRGB::Black;
+          }
+        }
+      }
+    }
+  }
+  else if (currentLEDState == FIRE)
+  {
+    int count = animation_step_count(fireUpdate, 2);
+    if (count > 0)
+    {
+      //int step = animation_step(fireUpdate, 2);
+      //random16_add_entropy( random(0)); //random() don't exists in ESP framework - workaround?
+
+      Fire2012WithPalette();
+    }
+  }
+  FastLED.show();
+}
+
+boolean FASTLEDtoMQTT()
+{
+  return false;
+}
+void MQTTtoFASTLEDJSON(char *topicOri, JsonObject&  jsonData)
+{
+  trc(F("MQTTtoFASTLEDJSON: "));
+  currentLEDState = GENERAL;
+  String topic = topicOri;
+  
+  trc(topic);
+  //number = (long)strtol(&datacallback[1], NULL, 16);
+
+  
+  
+  if (topic == subjectMQTTtoFASTLEDsetled)
+  {
+    
+    
+    
+   
+      trc(F("JSON parsed"));
+      int ledNr = jsonData["led"];
+      trc(F("led"));
+      trc(ledNr);
+      const char *color = jsonData["hex"];
+      trc(F("hex"));
+      
+      long number = (long)strtol(color, NULL, 16);
+      trc(number);
+      bool blink = jsonData["blink"];
+      if (ledNr <= FASTLED_NUM_LEDS)
+      {
+        trc(F("blink"));
+        trc(blink);
+        blinkLED[ledNr] = blink;
+        leds[ledNr] = number;
```

Comment:
```suggestion
        ledColorBlink[ledNr] = number;
```

Reply:
I updated the code

Reply:
there is a 50% chance that the color gets lost with the current code. You must also remove  line 104 ` ledColorBlink[i] = leds[i];`.

Reply:
At the moment I used the color "black" to turn LED off and I store the "current" color in the array ledColorBlink. When I remove this line the color could not be restored.

Reply:
i though the default is the off state and you set blink with the color it should have when it's on. the `number` contains the rgb value that should be the on state value.

Reply:
`leds `is the "color" array which is pointed to FastLED. `ledColorBlink `save the color while blinking. `number` is the rgb color which comes from the JSON.

Reply:
can you make an example of the used json and the order in which messages are sent to the topics

Reply:
I will leave it like it is.